src/network/network_crypto.cpp

```diff
@@ -0,0 +1,112 @@
+	std::array<uint8_t, 2 * size> keys;
+public:
+	~SharedKeys() { crypto_wipe(keys); }
+	constexpr std::span<const uint8_t> ClientToServer() const { return std::span(this->keys.data(), size); }
+	constexpr std::span<const uint8_t> ServerToClient() const { return std::span(this->keys.data() + size, size); }
+	constexpr std::span<uint8_t> Both() { return this->keys; }
+};
+
+X25519Key::~X25519Key()
+{
+	crypto_wipe(*this);
+}
+
+static bool X25519KeyExchange(SharedKeys<32> &shared_keys, const X25519Key &my_private_key, const X25519Key &their_public_key)
+{
+	X25519Key shared_secret;
+	crypto_x25519(shared_secret.data(), my_private_key.data(), their_public_key.data());
+	if (std::all_of(shared_secret.begin(), shared_secret.end(), [](auto v) { return v == 0; })) {
+		/* Secret is all 0 because public key is all 0, just reject it */
+		return false;
+	}
+
+	crypto_blake2b_ctx ctx;
+	crypto_blake2b_init(&ctx, shared_keys.Both().size());
+	crypto_blake2b_update(&ctx, shared_secret.data(), shared_secret.size());
+	crypto_blake2b_update(&ctx, my_public_key.data(), my_public_key.size());
+	crypto_blake2b_update(&ctx, their_public_key.data(), their_public_key.size());
+	crypto_blake2b_final(&ctx, shared_keys.Both().data());
+	return true;
+}
+
+
+X25519AuthorizedKeyHandler::X25519AuthorizedKeyHandler()
+{
+	RandomBytesWithFallback(this->my_private_key);
+	crypto_x25519_public_key(this->my_public_key.data(), this->my_private_key.data());
+
+	RandomBytesWithFallback(this->nonce);
+}
+
+X25519AuthorizedKeyHandler::~X25519AuthorizedKeyHandler()
+{
+	crypto_wipe(this->nonce);
+}
+
+/* virtual */ void X25519AuthorizedKeyHandler::SendRequest(Packet &p)
+{
+	p.Send_bytes(this->my_public_key);
+	p.Send_bytes(this->nonce);
+}
+
+/* virtual */ bool X25519AuthorizedKeyHandler::ReceiveResponse(Packet &p)
+{
+	X25519Key their_public_key{};
+	X25519Key message{};
+	std::array<uint8_t, 16> mac; // message authentication code
+
+	p.TransferOut(&TransferOutSpan, their_public_key);
+	p.TransferOut(&TransferOutSpan, message);
+	p.TransferOut(&TransferOutSpan, mac);
+
+	SharedKeys<32> shared_keys;
+	bool valid_exchange = X25519KeyExchange(shared_keys, this->my_private_key, their_public_key);
+	bool valid_unlock = crypto_aead_unlock(message.data(), mac.data(), shared_keys.ClientToServer().data(),
+			nonce.data(), their_public_key.data(), their_public_key.size(), message.data(), message.size()) == 0;
+	bool valid_message = crypto_verify32(message.data(), this->my_public_key.data());
+
+	const auto &allowed_public_keys = _settings_client.network.admin_public_keys;
+	std::string their_public_key_string = FormatArrayAsHex(their_public_key);
+	bool allowed_public_key = std::find(allowed_public_keys.begin(), allowed_public_keys.end(), their_public_key_string) != allowed_public_keys.end();
+
+	/* Check everything at the end, as to attempt to not leak much information about what went wrong via timing. */
+	return valid_exchange && valid_unlock && valid_message && allowed_public_key;
```

Comment:
I am not sure about this. I get where you are coming from, but it is rather unusual. And I wouldn't worry about timing attacks on this level as much.

The issue you get now, is that `shared_keys` might be zeros, which you feed to `crypto_aead_unlock`, which will have different behaviour. So your timing is influenced anyway.

I am more tempted to follow monocypher's example, and just return when things return an error. It just makes things less complicated.